File: frontend/src/types/execution.ts
export interface NodeExecutionLog {
  node_id: string
  node_type: string
  status: string
  inputs?: Record<string, unknown> | null
  outputs?: Record<string, unknown> | null
  error?: string | null
  duration_ms?: number | null
  started_at?: string | null
  completed_at?: string | null
}

export interface Execution {
  id: string
  workflow_id: string | null
  workspace_id: string
  workflow_version?: number | null
  status: string
  trigger_type?: string | null
  input_data?: Record<string, unknown> | null
  output_data?: Record<string, unknown> | null
  error_message?: string | null
  execution_trace?: { nodes: NodeExecutionLog[] } | null
  execution_time_ms?: number | null
  token_usage?: { total_tokens?: number; prompt_tokens?: number; completion_tokens?: number } | null
  cost_usd?: number | null
  started_at?: string | null
  completed_at?: string | null
  created_at: string
}

export interface ExecutionListResponse {
  executions: Execution[]
  total: number
}

File: backend/app/schemas/execution.py
from pydantic import BaseModel
from datetime import datetime
from typing import Any


class ExecuteRequest(BaseModel):
    input_data: dict[str, Any] = {}
    async_execution: bool = False


class NodeExecutionLog(BaseModel):
    node_id: str
    node_type: str
    status: str
    inputs: dict[str, Any] | None = None
    outputs: dict[str, Any] | None = None
    error: str | None = None
    duration_ms: int | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None


class ExecutionResponse(BaseModel):
    id: str
    workflow_id: str | None = None
    workspace_id: str
    workflow_version: int | None = None
    status: str
    trigger_type: str | None = None
    input_data: dict | None = None
    output_data: dict | None = None
    error_message: str | None = None
    execution_trace: dict | None = None
    execution_time_ms: int | None = None
    token_usage: dict | None = None
    cost_usd: float | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ExecutionListResponse(BaseModel):
    executions: list[ExecutionResponse]
    total: int
